src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1042,83 +1044,102 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
-                                // an RFQ. Instead, the server closes the connection immediately
-                                throw error;
-                            }
+                                continue;
 
-                            continue;
+                            case BackendMessageCode.ReadyForQuery:
+                                if (error != null)
+                                {
+                                    NpgsqlEventSource.Log.CommandFailed();
+                                    throw error;
+                                }
+                                // Just in case if we were successful in sending a cancellation, but the query is already completed
+                                originalTimeoutException = null;
 
-                        case BackendMessageCode.ReadyForQuery:
-                            if (error != null)
-                            {
-                                NpgsqlEventSource.Log.CommandFailed();
-                                throw error;
-                            }
+                                break;
 
-                            break;
+                            // Asynchronous messages which can come anytime, they have already been handled
+                            // in ParseServerMessage. Read the next message.
+                            case BackendMessageCode.NoticeResponse:
+                            case BackendMessageCode.NotificationResponse:
+                            case BackendMessageCode.ParameterStatus:
+                                Debug.Assert(msg == null);
+                                if (!readingNotifications2)
+                                    continue;
+                                return null;
+                            }
 
-                        // Asynchronous messages which can come anytime, they have already been handled
-                        // in ParseServerMessage. Read the next message.
-                        case BackendMessageCode.NoticeResponse:
-                        case BackendMessageCode.NotificationResponse:
-                        case BackendMessageCode.ParameterStatus:
-                            Debug.Assert(msg == null);
-                            if (!readingNotifications2)
-                                continue;
-                            return null;
+                            Debug.Assert(msg != null, "Message is null for code: " + messageCode);
+                            return msg;
+                        }
+                        catch (NpgsqlException e) when (!readingNotifications2 && e.InnerException is TimeoutException && originalTimeoutException is null)
+                        {
+                            // We have got a timeout while not reading the async notifications - trying to cancel a query
+                            try
+                            {
+                                CancelRequest(true);
```

Comment:
If I'm following the code correctly, this means we wait for CommandTimeout once, do a cancellation request, and then wait CommandTimeout again before breaking the connection. Maybe this is OK, but I think doubling the timeout until the code is unblocked (in case of a network partition) may be  too much. We may want to make the second timeout shorter, or at least allow the user to control it.

@YohDeadfall @Brar what do you think about this?

Reply:
I agree that doubling the timeout is maybe a bit long and probably unexpected.
What I'd expect probably would depend on the actual timeout I set.
If I'd set it to one second I probably wouldn't care a lot if it took two 1300ms but if I'd set it to one minute 80s would probably be too long (or even 80 minutes in case of an hour).
So IMHO we have to options here.

1. Make the delay configurable
   This is certainly the most flexible solution but to me it looks like over-engineering the problem. I'd rather not do it unless we get a profound use case.
2. Add a relative amount (let's say 30%) but no more than a certain amount (let's say two seconds).
   This is what would probably be the least surprising solution for people who don't know and care about those details.

Reply:
Wouldn't the relative amount only work up to 6 seconds timeouts (30% from 6 is 1.8 seconds)? I think, we should just set 2 seconds for now.

Reply:
Added as a temporary solution a 2 seconds timeout

Reply:
I'm not too sure about the percentage approach here. The user-provided CommandTimeout is usually less about networking delays, and more about how much queries can take before returning a response (this isn't absolute for sure). A query that's very heavy processing-wise can time out even against localhost. On the other hand, PG cancellation doesn't depend on the query itself - as long as the network is working well and PG is in a good state, it should be instantaneous.

So while the first timeout, in my mind, tends to target slow queries and/or deadlocks, the second one is purely about networking (delays or partition).

I vote for adding a connection string parameter and defaulting to something reasonable. 2 seconds is probably reasonable for a local network, but not necessarily for a cloud PG running across the world... Adding the parameter itself is trivial, and I don't see much downside to it (except maybe another knob that could confuse users in the docs).

Reply:
> I vote for adding a connection string parameter and defaulting to something reasonable.

You're the boss 😀 . I'll take a look a little bit later.